src/lib/statistics/column_statistics.cpp

```diff
@@ -26,52 +26,55 @@ std::shared_ptr<BaseColumnStatistics> ColumnStatistics<ColumnDataType>::clone()
 template <typename ColumnDataType>
 FilterByValueEstimate ColumnStatistics<ColumnDataType>::estimate_predicate_with_value(
     const PredicateCondition predicate_condition, const AllTypeVariant& variant_value,
-    const std::optional<AllTypeVariant>& value2) const {
-  const auto value = type_cast_variant<ColumnDataType>(variant_value);
+    const std::optional<AllTypeVariant>& variant_value2) const {
+  const auto value = lossless_variant_cast<ColumnDataType>(variant_value);
+  if (!value) return {non_null_value_ratio(), without_null_values()};
```

Comment:
Does this mean that all comparisons that use different types will have no statistics? Should we care if we lose the double precision here?

Reply:
One of the rougher edges of this PR, I agree. But I really wanted to remove type_cast and the bug proneness that it brings and was ready to sacrifice estimations on non-equal operand types.

There is a (potential) follow-up to this PR: `std::pair<PredicateCondition, Target> predicate_operand_cast<Target>(Source, PredicateCondition)` (exact name to be determined). This will take, e.g., a predicate `<int_column> > 5.5` and turn it into `<int_column> > 5` or `<int_column> >= 5.3` into `<int_column> >= 6`.

Reply:
With the number of open issues (and the histogram PR touching on this as well), I am not sure if we should focus on that or wait for that. The statistics will always be a best-effort thing, and accepting type conversion errors here will certainly be better than not returning any statistics at all.

Reply:
Maybe we don't need all back-and-forths that the old type_cast had, but if static_cast can deal with it, we should be able to use it.

Reply:
Normally I am all for merging PRs fast and opening follow-up issues to keep my pipeline small.

But cases where actual functionality is (temporarily) removed by a PR shouldn't happen. I introduced `static_variant_cast` as a local helper for the statistics. If I need it in #1489 I'll refactor it into its own file.